stmhal/dma.c

```diff
@@ -148,75 +342,88 @@ static void dma_enable_clock(int dma_id) {
             // in dma_last_channel (for DMA1) needs to be invalidated.
 
             for (int channel = NSTREAMS_PER_CONTROLLER; channel < NSTREAM; channel++) {
-                dma_last_channel[channel] = DMA_INVALID_CHANNEL;
+                dma_last_sub_instance[channel] = DMA_INVALID_CHANNEL;
             }
         }
     }
 }
 
-static void dma_disable_clock(int dma_id) {
+static void dma_disable_clock(dma_id_t dma_id) {
     // We just mark the clock as disabled here, but we don't actually disable it.
     // We wait for the timer to expire first, which means that back-to-back
     // transfers don't have to initialize as much.
     dma_tickle(dma_id);
     dma_enable_mask &= ~(1 << dma_id);
 }
 
-void dma_init(DMA_HandleTypeDef *dma, DMA_Stream_TypeDef *dma_stream, const DMA_InitTypeDef *dma_init, uint32_t dma_channel, uint32_t direction, void *data) {
-    int dma_id = get_dma_id(dma_stream);
-    //printf("dma_init(%p, %p(%d), 0x%x, 0x%x, %p)\n", dma, dma_stream, dma_id, (uint)dma_channel, (uint)direction, data);
+void dma_init_handle(DMA_HandleTypeDef *dma, const dma_descr_t * dma_descr, void *data) {
+    if (dma_descr != NULL) {
+        // initialise parameters
+        dma->Instance = dma_descr->instance;
+        dma->Init = *dma_descr->init;
+        dma->Init.Direction = dma_descr->transfer_direction;
+#if defined(MCU_SERIES_L4)
+        dma->Init.Request = dma_descr->sub_instance;
+#else
+        dma->Init.Channel = dma_descr->sub_instance;
+#endif
+        // half of __HAL_LINKDMA(data, xxx, *dma)
+        // caller must implement other half by doing: data->xxx = dma
+        dma->Parent = data;
+    }
+
+    return;
+}
 
+void dma_init(DMA_HandleTypeDef *dma, const dma_descr_t * dma_descr, void *data){
     // Some drivers allocate the DMA_HandleTypeDef from the stack
     // (i.e. dac, i2c, spi) and for those cases we need to clear the
     // structure so we don't get random values from the stack)
     memset(dma, 0, sizeof(*dma));
 
-    // set global pointer for IRQ handler
-    dma_handle[dma_id] = dma;
 
-    // initialise parameters
-    dma->Instance = dma_stream;
-    dma->Init = *dma_init;
-    dma->Init.Direction = direction;
-    dma->Init.Channel = dma_channel;
```

Comment:
Moved the last 4 lines to `dma_init_handle` to encapsulate this low level initialization.